### ISSUE TITLE
Is it possible to reuse animation for multiple objects in the scene as per the GLTF specification ?

### ISSUE THREAD

**Issue by anuprao** [KhronosGroup/glTF#933]
Hi,

```
"animation_0": {
            "channels": [
                {
                    "sampler": "animation_0_scale_sampler",
                    "target": {
                        "id": "node-cam01-box",
                        "path": "scale"
                    }
                },
                {
                    "sampler": "animation_0_translation_sampler",
                    "target": {
                        "id": "node-cam01-box",
                        "path": "translation"
                    }
                },
                {
                    "sampler": "animation_0_rotation_sampler",
                    "target": {
                        "id": "node-cam01-box",
                        "path": "rotation"
                    }
                }
            ],
            "parameters": {
                "TIME": "animAccessor_0",
                "scale": "animAccessor_1",
...

```

As per curent specification for animations, an **id** field in the **target** entry under **channels** describes which object must be setup for this channel of the animation. How can the animation be reused for a large number (say 1000) of objects in the scene ? Does the author have to append 1000 channels each with distinct **id** ? , or, is it possible to specify a list of IDs in the **id** field ?

A simple use case would be an animated 'button press like' animation which might need to be applied to several 'button-like' instances in a scene or multiple scenes.

Regards,
Anup Jayapal Rao

**Comment by pjcozzi** [comment]
> Does the author have to append 1000 channels each with distinct id ?

I believe yes.

> is it possible to specify a list of IDs in the id field ?

If there are enough use cases, this could be considered for the future.

**Comment by donmccurdy** [comment]
I'd be inclined against adding this — the animation keyframes don't need to be duplicated anyway (the accessors can be reused) and this seems like a problem more commonly solved by animating a parent node, which would also be easier for the engine to optimize than 1000 discrete but identical animations.

tl;dr, it's an optimization likely to encourage people to do something less efficient.

The "multiple scenes" case may be valid, but sounds like an application-specific issue that could be better addressed with vendor extensions initially, e.g. to define animations that are reusable or even transmitted separately from the model.

**Comment by pjcozzi** [comment]
If folks agree with @donmccurdy, let's close this.  Please chime in!

**Comment by anuprao** [comment]
I agree with @donmccurdy with respect to reusing animation data is my preferred choice as well. Promoting an efficient way to do things is always a good practice. 

The example use case I provided was to indicate a common problem in GUIs. Multiple buttons might contain similar click/press/release animation. In such cases, The animation for such events cannot be applied to the parent dialog( or form) containing the buttons as suggested above. That would affect the entire dialog or form. The animation would have to be applied only to buttons relevant to the event. However, in current scheme of things, this might lead to duplication of animation data.

Hence my case, and why I would like to refer to the same animation data for all objects behaving identically. (This in some way is identical to reusing the same CSS animations/transistions def for all buttons that share the same aspect.)

Please correct me if my assumptions are incorrect in any way.

**Comment by donmccurdy** [comment]
Yeah, your assumptions are reasonable. But there's not a best practice way to export GUIs, at least that I know of, and tools like Blender or Maya aren't going to export things in the way you're describing here. So it would be pretty risky for us to define a data structure that the common tooling the JSON descriptions of animations is a fairly minor thing compared to duplicating the underlying keyframes, and glTF does already help with that.

So given that the tools for creating a GUI and the loaders to run them are likely to be custom built, it would be better for any modifications to glTF to be done via [vendor extensions](https://github.com/KhronosGroup/glTF/tree/master/extensions#extension-mechanics). If it turns out to be broadly useful, we can always promote those to official extensions or part of the core glTF specification.

Going to close this issue, but certainly feel free to experiment or propose an extension!

**Comment by anuprao** [comment]
@donmccurdy 
You mentioned  _duplicating JSON wrappers around animations is a fairly minor thing compared to duplicating the underlying keyframe accessors, and glTF does already handle that._

Kindly elaborate on this with a simple example.

This will help those like me to tackle similar scenarios.

**Comment by donmccurdy** [comment]
Sure, here's a snippet of an `animations` entry:

```json
"animations" : [
   {
      "channels" : [
         {
            "sampler" : 0,
            "target" : {
               "node" : 0,
               "path" : "rotation"
            }
         },
         {
            "sampler" : 0,
            "target" : {
               "node" : 1,
               "path" : "rotation"
            }
         }
      ],
      "name" : "animation_AnimatedMultiCube",
      "samplers" : [
         {
            "input" : 0,
            "interpolation" : "LINEAR",
            "output" : 1
         }
      ]
   }
],
```

Note that both channels point to the same sampler. It would also be valid to have multiple samplers pointing to the same `input` or `output` accessors — inputs especially can probably be reused efficiently across many channels.

Here's the working result:

[AnimatedMultiCube.zip](https://github.com/KhronosGroup/glTF/files/1591956/AnimatedMultiCube.zip)

I'm less familiar with how this would be implemented within the Blender exporter, I think you'd want to test that and report issues there.

**Comment by anuprao** [comment]
Thanks a ton @donmccurdy 
Appreciate your inputs.

I will publish my findings after experimenting with your suggestions.
